Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,534 @@
+// Copyright (c) 2021 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+#if SWIFTNIO_USE_IO_URING
+#if os(Linux)
+
+import CNIOLinux
+
+@usableFromInline
+enum CQEEventType : UInt8 {
+    case poll = 1, pollModify, pollDelete // start with 1 to not get zero bit patterns for stdin
+}
+
+internal enum UringError: Error {
+    case loadFailure
+    case uringSetupFailure
+    case uringWaitCqeFailure
+}
+
+internal extension TimeAmount {
+    func kernelTimespec() -> __kernel_timespec {
+        var ts = __kernel_timespec()
+        ts.tv_sec = self.nanoseconds / 1_000_000_000
+        ts.tv_nsec = self.nanoseconds % 1_000_000_000
+        return ts
+    }
+}
+
+// URingUserData supports (un)packing into an `UInt64` as io_uring has a user_data 64-bit payload which is set in the SQE
+// and returned in the CQE. We're using 56 of those 64 bits, 32 for the file descriptor, 16 for a "registration ID" and 8
+// for the type of event issued (poll/modify/delete).
+@usableFromInline struct URingUserData {
+    @usableFromInline var fileDescriptor: CInt
+    @usableFromInline var registrationID: UInt16 // SelectorRegistrationID truncated, only have room for bottom 16 bits (could be expanded to 24 if required)
+    @usableFromInline var eventType: CQEEventType
+    @usableFromInline var padding: Int8 // reserved for future use
+
+    @inlinable init(registrationID: SelectorRegistrationID, fileDescriptor: CInt, eventType: CQEEventType) {
+        assert(MemoryLayout<UInt64>.size == MemoryLayout<URingUserData>.size)
+        self.registrationID = UInt16(truncatingIfNeeded: registrationID.rawValue)
+        self.fileDescriptor = fileDescriptor
+        self.eventType = eventType
+        self.padding = 0
+    }
+
+    @inlinable init(rawValue: UInt64) {
+        let unpacked = IntegerBitPacking.unpackUInt32UInt16UInt8(rawValue)
+        self = .init(registrationID: SelectorRegistrationID(rawValue: UInt32(unpacked.1)),
+                     fileDescriptor: CInt(unpacked.0),
+                     eventType: CQEEventType(rawValue:unpacked.2)!)
+    }
+}
+
+extension UInt64 {
+    init(_ uringUserData: URingUserData) {
+        let fd = uringUserData.fileDescriptor
+        let eventType = uringUserData.eventType.rawValue
+        assert(fd >= 0, "\(fd) is not a valid file descriptor")
+        assert(eventType >= 0, "\(eventType) is not a valid eventType")
+
+        self = IntegerBitPacking.packUInt32UInt16UInt8(UInt32(truncatingIfNeeded: fd),
+                                                       uringUserData.registrationID,
+                                                       eventType)
+    }
+}
+
+// These are the events returned up to the selector
+internal struct UringEvent {
```

Comment:
nit: elsewhere you use `URing` (not `Uring` like here). I think we should be consistent and use `URing`

Reply:
Fixed.